contracts/halo-pool/src/contract.rs

```diff
@@ -556,77 +1117,138 @@ pub fn query(deps: Deps, env: Env, msg: QueryMsg) -> Result<Binary, ContractErro
             Ok(to_binary(&query_pending_reward(deps, env, address)?)?)
         }
         QueryMsg::TotalStaked {} => Ok(to_binary(&query_total_lp_token_staked(deps, env)?)?),
+        QueryMsg::StakerInfo { address } => Ok(to_binary(&query_staker_info(deps, address)?)?),
     }
 }
 
-fn query_pool_info(deps: Deps) -> Result<PoolInfo, ContractError> {
-    let pool_info: PoolInfo = POOL_INFO.load(deps.storage)?;
-    let res = PoolInfo { ..pool_info };
-    Ok(res)
+fn query_pool_info(deps: Deps) -> Result<PoolInfos, ContractError> {
+    Ok(POOL_INFOS.load(deps.storage)?)
 }
 
 fn query_pending_reward(
     deps: Deps,
     env: Env,
     address: String,
-) -> Result<RewardTokenAsset, ContractError> {
+) -> Result<RewardTokenAssetResponse, ContractError> {
     // Get current time
-    let current_time = env.block.time;
+    let current_time = env.block.time.seconds();
+    // Get pool infos
+    let pool_infos = POOL_INFOS.load(deps.storage)?;
+    // Check if staker has staked in the pool
+    if STAKERS_INFO
+        .may_load(deps.storage, Addr::unchecked(address.clone()))?
+        .is_none()
+    {
+        return Ok(RewardTokenAssetResponse {
+            info: pool_infos.reward_token,
+            amount: Uint128::zero(),
+            time_query: current_time,
+        });
+    }
+    // Get current pool index
+    let current_phase_index = pool_infos.current_phase_index;
+    // Get current pool info in pool infos
+    let pool_info = POOL_INFOS.load(deps.storage)?.pool_infos[current_phase_index as usize].clone();
     // Get last reward time
-    let last_reward_time = LAST_REWARD_TIME.load(deps.storage)?;
+    let last_reward_time = pool_infos.last_reward_time;
     // Get accrued token per share
-    let mut accrued_token_per_share = ACCRUED_TOKEN_PER_SHARE.load(deps.storage)?;
-    // Get pool info
-    let pool_info: PoolInfo = POOL_INFO.load(deps.storage)?;
+    let mut accrued_token_per_share = pool_infos.accrued_token_per_share;
+    // Multiply factor
+    let mut multiplier: u64;
+
     // Get staker info
     let staker_info = STAKERS_INFO
-        .may_load(deps.storage, Addr::unchecked(address))?
-        .unwrap_or(StakerRewardAssetInfo {
-            amount: Uint128::zero(),
-            reward_debt: Uint128::zero(),
-        });
+        .load(deps.storage, Addr::unchecked(address))
+        .unwrap();
+
+    // get staker joined phases
+    let staker_joined_phase = staker_info.joined_phase;
+    // Init reward amount
+    let mut reward_amount = Uint128::zero();
+
+    if staker_joined_phase < current_phase_index {
+        for i in staker_joined_phase..current_phase_index {
```

Comment:
Bỏ if đi được mà. for thì để là staker_joined_phase..current_phase_index-1

Reply:
Removed